graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IncrementModelBuilder.java

```diff
@@ -0,0 +1,146 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.stargate.graphql.schema.graphqlfirst.processor;
+
+import graphql.Scalars;
+import graphql.language.Directive;
+import graphql.language.InputValueDefinition;
+import graphql.language.ListType;
+import graphql.language.Type;
+import graphql.language.TypeName;
+import graphql.schema.GraphQLScalarType;
+import io.stargate.graphql.schema.graphqlfirst.util.TypeHelper;
+import io.stargate.graphql.schema.scalars.CqlScalar;
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class IncrementModelBuilder extends ModelBuilderBase<IncrementModel> {
+
+  public static final String COUNTER_TYPE_NAME = "counter";
+  private final EntityModel entity;
+  private final FieldModel field;
+  private final Map<String, EntityModel> entities;
+  private final ProcessingContext context;
+
+  private static final boolean PREPEND_DEFAULT = false;
+  private final String operationName;
+  private final InputValueDefinition argument;
+
+  public IncrementModelBuilder(
+      InputValueDefinition argument,
+      String operationName,
+      EntityModel entity,
+      FieldModel field,
+      Map<String, EntityModel> entities,
+      ProcessingContext context) {
+    super(context, argument.getSourceLocation());
+    this.argument = argument;
+    this.operationName = operationName;
+    this.entity = entity;
+    this.field = field;
+    this.entities = entities;
+    this.context = context;
+  }
+
+  @Override
+  IncrementModel build() throws SkipException {
+
+    Optional<Directive> directive = DirectiveHelper.getDirective(CqlDirectives.INCREMENT, argument);
+
+    boolean prepend =
+        directive
+            .flatMap(d -> DirectiveHelper.getBooleanArgument(d, "prepend", context))
+            .orElse(PREPEND_DEFAULT);
+
+    validate(field, prepend);
+    return new IncrementModel(field, prepend, argument.getName());
+  }
+
+  protected void validate(FieldModel field, boolean prepend) throws SkipException {
+    if (field.isPartitionKey() || field.isClusteringColumn()) {
+      invalidMapping(
+          "Operation %s: directive %s is not supported for partition/clustering key field %s.",
+          operationName, CqlDirectives.INCREMENT, field.getGraphqlName());
+      throw SkipException.INSTANCE;
+    } else {
+      checkValidForRegularColumn(field, prepend);
+    }
+  }
+
+  private void checkValidForRegularColumn(FieldModel field, boolean prepend) throws SkipException {
+
+    Type<?> fieldInputType =
+        toInput(field.getGraphqlType(), argument, entity, field, entities, operationName);
+    // if it is non-list type
+    if (fieldInputType instanceof TypeName) {
+      validateCounter((TypeName) fieldInputType);
+      validatePrepend(field, prepend);
+    } else {
+      // otherwise it must be a list
+      checkArgumentIsListOf(argument, entity, entities, operationName, field);
+    }
```

Comment:
We should also fail if `prepend` is set and the column is a a CQL set.

Reply:
Is it even possible to create a set using the GrapqhQL API?
I don't see any test covering that case.

Reply:
The CQL type is known at this stage, via `field.getCqlType()`. Fixed in 3e2b33f0.